### PR TITLE
Generic email setup and basic notification setup

### PR DESCRIPTION
**Pull request by Avi-Rana-1718**

### CODE REVIEW

File: src/main/resources/db/migration/V1__create__notification__table.sql [hunk 1]
```diff
@@ -0,0 +1,11 @@
+CREATE TYPE notificationType AS ENUM ('email', 'sms', 'whatsapp');
+CREATE TYPE notificationStatus AS ENUM('delivered', 'failed', 'requested', 'unknown');
+
+CREATE TABLE notification (
+    notificationId UUID PRIMARY KEY,
+    userId UUID NOT NULL,
+    triggerDate TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,
+    type notificationType NOT NULL,
+    status notificationStatus NOT NULL,
+    retryCount INTEGER DEFAULT 0
```
[the remark this answers is not in the document]
Comment (Avi-Rana-1718) [reply src/main/resources/db/migration/V1__create__notification__table.sql]: Fixed!

**Comment by Avi-Rana-1718** [comment]
@solangelo101 Please provide suggestions for writing test.

File: src/main/java/com/notification/herald/utils/MailUtil.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+    public String sendMail(MailRequestDto request, MailProviderEnum provider) throws Exception {
+        String providerName = provider.getValue().toLowerCase();
+        MailProvider mailProvider = providerList.get(providerName);
+
+        if(Objects.isNull(mailProvider)) {
+            throw new BadRequestException("Invalid mail provider!");
+        }
+
+        return mailProvider.sendMail(request).block();
```
Comment (Avi-Rana-1718) [inline src/main/java/com/notification/herald/utils/MailUtil.java]: Will look into this.

File: src/main/java/com/notification/herald/dto/mail/Mailjet/request/MailjetRequestDto.java [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package com.notification.herald.dto.mail.Mailjet.request;
+
+import lombok.Builder;
+
+import java.util.List;
+
+@Builder
```
[the remark this answers is not in the document]
Comment (solangelo101) [reply src/main/java/com/notification/herald/dto/mail/Mailjet/request/MailjetRequestDto.java]: Why was this resolved? Builder isn't required in this scenario.

File: src/main/java/com/notification/herald/dto/ResponseDto.java [hunk 4]
```diff
@@ -1,6 +1,5 @@
 package com.notification.herald.dto;
 
 public record ResponseDto(
-    Object data,
-    boolean success
+    Object data
```
Comment (solangelo101) [inline src/main/java/com/notification/herald/dto/ResponseDto.java]: Can we add a status code to this as well ?

File: src/main/java/com/notification/herald/entities/NotificationEntity.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
+package com.notification.herald.entities;
+
+import com.notification.herald.enums.NotificationStatusEnum;
+import com.notification.herald.enums.NotificationTypeEnum;
+import jakarta.persistence.*;
```
Comment (solangelo101) [inline src/main/java/com/notification/herald/entities/NotificationEntity.java]: Don't use * imports.

**Review by solangelo101 — CHANGES_REQUESTED** [review]
Can we add unit testing for this? I saw that we tested a success sceanrio but we don't have any testing done for negative scenarios.
Also add a README that shows how to send a sample request.

File: src/main/java/com/notification/herald/dto/mail/Mailjet/request/MailjetRequestDto.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Avi-Rana-1718) [reply src/main/java/com/notification/herald/dto/mail/Mailjet/request/MailjetRequestDto.java]: Removed builder

File: src/main/java/com/notification/herald/dto/ResponseDto.java (hunk 4 above)
Comment (Avi-Rana-1718) [reply src/main/java/com/notification/herald/dto/ResponseDto.java]: Added status code but did it manual way, maybe we can make a interceptor for it.

File: src/main/java/com/notification/herald/dto/mail/MailAddress.java [hunk 6]
```diff
@@ -0,0 +1,10 @@
+package com.notification.herald.dto.mail;
+
+import jakarta.validation.constraints.NotNull;
+
+public record MailAddress(
+        @NotNull
```
[the remark this answers is not in the document]
Comment (Avi-Rana-1718) [reply src/main/java/com/notification/herald/dto/mail/MailAddress.java:7]: Added email validation

File: src/main/java/com/notification/herald/entities/NotificationEntity.java (hunk 5 above)
Comment (Avi-Rana-1718) [reply src/main/java/com/notification/herald/entities/NotificationEntity.java]: Fixed
